File: src/MatchResult.cc
#include <algorithm>

#include "MatchResult.h"
#include "dp.h"
#include "ContigMapData.h"
#include "OpticalMapData.h"

#include <iostream>

#define MR_DEBUG 0

void MatchResult::setMatchedChunks(const MatchedChunkVec& matchedChunkList)
{
    matchedChunkList_ = matchedChunkList;
    buildAlignmentAttributes();
}

ostream& operator<<(ostream& os, const MatchResult& mr)
{
    os << mr.contigId_ << " "
       << mr.chromosomeId_ << " "
       << mr.contigLength_ << " "
       << mr.forward_ << " "
       << mr.opStartIndex_ << " "
       << mr.opEndIndex_ << " "
       << mr.cStartIndex_ << " "
       << mr.cEndIndex_ << " "
       << mr.totalMisses_ << " ";
       //<< mr.pval_;
    return os;
}


void MatchResult::buildAlignmentAttributes()
{

    if (matchedChunkList_.empty()) return;

    contigUnalignedFrags_ = 0;
    opticalMisses_ = 0;
    contigMisses_ = 0;
    contigHits_  = 0;
    opticalHits_ = 0;
    opticalTotalAlignedBases_ = 0;
    contigTotalAlignedBases_ = 0;
    contigInnerAlignedBases_ = 0;
    opticalInnerAlignedBases_ = 0;
    contigUnalignedBases_ = 0;
    numAlignedInnerBlocks_ = 0;

    // Assign the starting and ending bp locations.
    const MatchedChunk& firstChunk = matchedChunkList_.front();
    opStartIndex_ = firstChunk.getOpticalStartIndex();
    cStartIndex_ = firstChunk.getContigStartIndex();
    opStartBp_ = firstChunk.getOpticalStartBp();
    cStartBp_ = firstChunk.getContigStartBp();


    const MatchedChunk& lastChunk = matchedChunkList_.back();
    opEndIndex_ = lastChunk.getOpticalEndIndex() - 1; // inclusive
    cEndIndex_ = lastChunk.getContigEndIndex() - 1; // inclusive
    opEndBp_ = lastChunk.getOpticalEndBp();
    cEndBp_ = lastChunk.getContigEndBp();

    // Loop over the vector of matched fragment chunks and
    // compute alignment statistics and descriptive strings.
    typedef vector<MatchedChunk>::const_iterator ChunkIter;
    const ChunkIter mb = matchedChunkList_.begin();
    const ChunkIter me = matchedChunkList_.end();


    for (ChunkIter mi = mb; mi != me; mi++)
    {
        bool firstAlignment = (mi == mb);
        int cl = mi->getContigMatchLengthBp();
        int ol = mi->getOpticalMatchLengthBp();

        if (mi->isContigGap())
        {
            contigUnalignedFrags_ += mi->getNumContigFrags();
            contigUnalignedBases_ += cl;
            continue;
        } 

        opticalMisses_ += mi->getNumOpticalMisses();
        contigMisses_ += mi->getNumContigMisses();
        opticalTotalAlignedBases_ += ol;
        contigTotalAlignedBases_ += cl;
        if (!mi->isBoundaryChunk())
        {
            // Model assumes that Optical Frag ~ N(C, SIGMA^2*C) where C is contig frag size
            double var = Constants::SIGMA2 * cl;
            double d = ol - cl;
            chi2_ += d*d/var;
            contigInnerAlignedBases_ += cl;
            opticalInnerAlignedBases_ += ol;
            numAlignedInnerBlocks_++;
        }

        if (firstAlignment && !mi->isBoundaryChunk())
        {
            // The first alignment is not a boundary chunk.
            // Therefore, it is bounded by two restriction sites.
            // Add the contribution of the left matched sites to the hit count
            contigHits_ += 1;
            opticalHits_ += 1;
        }

        contigHits_ += 1;
        opticalHits_ += 1;
    }
    
    totalHits_ = contigHits_ + opticalHits_;
    totalMisses_ = (contigMisses_ + opticalMisses_);
    totalMissRate_ = (totalMisses_ == 0) ? 0.0 : 1.0*totalMisses_/(totalHits_ + totalMisses_);
    contigMissRate_ = (contigMisses_== 0) ? 0.0 : 1.0*contigMisses_/(contigHits_ + contigMisses_);
    opticalMissRate_ = (opticalMisses_ == 0) ? 0.0 : 1.0*opticalMisses_/(opticalHits_ + opticalMisses_);
    alignedLengthRatio_ = 1.0*min(contigInnerAlignedBases_, opticalInnerAlignedBases_) /
                              max(contigInnerAlignedBases_, opticalInnerAlignedBases_);
    contigUnalignedBaseRatio_ = 1.0*contigUnalignedBases_/contigLength_;

}

// Populate the descriptive strings for the alignment
void MatchResult::annotate()
{
    bool firstAlignedFrag = true;
    bool lastChunk = false;
    stringstream opt_ss, c_ss; // Strings representing the overall alignment
    stringstream opt_aligned, c_aligned; // Strings representing indices of aligned fragments
    stringstream c_lost; // String representing the lost fragment indices
    stringstream score_ss; score_ss.precision(3);

    // Loop over the vector of matched fragment chunks and
    // compute alignment statistics and descriptive strings.
    vector<MatchedChunk>::const_iterator mi,mb,me;
    mb = matchedChunkList_.begin();
    me = matchedChunkList_.end();
    for (mi = mb; mi != me; mi++)
    {
        lastChunk = (mi == me-1);

        if (mi->isContigGap())
        {
            opt_ss <<  " --- "; // Print a gap in the optical sequence
            c_ss << " (" << mi->getContigMatchLengthBp() << ") ";
            c_lost << (c_lost.str().size() > 0 ? "," : "") << mi->getContigStartIndex();

        } else {

            vector<FragData>::const_iterator it,itb,ite;

            if (firstAlignedFrag && !mi->isBoundaryChunk())
            {
                // This chunk is bounded by a restriction site
                // Semi-colon denotes a matched restriction site
                opt_ss << "; ";
                opt_aligned << "; ";
                c_ss << "; ";
                c_aligned << "; ";
            }


            // Add to alignment descriptive strings
            const vector<FragData>::const_iterator opE = mi->getOpticalFragE();
            const vector<FragData>::const_iterator cE = mi->getContigFragE();
            for (it = mi->getOpticalFragB();
                 it != opE;
                 it++)
                opt_ss << " " <<  it->size_ << "," << it->sd_ << " ";

            int lasti = mi->getOpticalEndIndex()-1;
            for (int i = mi->getOpticalStartIndex();
                 i <= lasti;
                 i++)
                opt_aligned << i << ( i==lasti  ? "" : ",");

            for (it = mi->getContigFragB();
                 it != cE;
                 it++)
                c_ss << " " <<  it->size_ << "," << " ";

            lasti = mi->getContigEndIndex()-1;
            for (int i = mi->getContigStartIndex(); i <= lasti; i++)
                c_aligned << i << ( i==lasti  ? "" : ",");

            // We post a trailing semi-colon denoting a matched site
            // if it is a any chunk other than the first
            if (firstAlignedFrag || !mi->isBoundaryChunk())
            {
                // Semi-colon denotes a matched restriction site
                c_ss << "; ";
                c_aligned << ";";
                opt_ss << "; ";
                opt_aligned << ";";
            }

            firstAlignedFrag = false;
        }

        const Score& score = mi->getScore();
        score_ss << "(" << score.contig << "," << score.optical << "," << score.sizing << ")";
        if (!lastChunk)
            score_ss << " / ";
    }


    contigAlignedIndexString_ = c_aligned.str();
    contigMatchString_ = c_ss.str();
    contigLostIndexString_ = c_lost.str();
    opticalAlignedIndexString_ = opt_aligned.str();
    opticalMatchString_ = opt_ss.str();
    scoreString_ = score_ss.str();
}


// Reset all attributes
void MatchResult::reset()
{

    // Contig information
    contigId_.clear();

    // Alignment location & orientation
    chromosomeId_.clear();
    opStartIndex_ = 0;
    opEndIndex_ = 0;
    opStartBp_ = 0;
    opEndBp_ = 0;
    forward_ = 0;
    matchedChunkList_.clear();

    // Alignment statistics
    score_ = 0;
    totalMisses_ = 0;
    totalHits_ = 0;
    totalMissRate_ = 0;
    numAlignedInnerBlocks_ = 0;
    contigMisses_ = 0;
    contigHits_ = 0;
    contigMissRate_ = 0;
    contigUnalignedFrags_ = 0;
    contigUnalignedBases_ = 0;
    contigTotalAlignedBases_ = 0;
    contigInnerAlignedBases_ = 0;
    opticalMisses_ = 0;
    opticalHits_ = 0;
    opticalMissRate_ = 0;
    opticalTotalAlignedBases_ = 0;
    opticalInnerAlignedBases_ = 0;
    alignedLengthRatio_ = 0;
    contigUnalignedBaseRatio_ = 0;
    chi2_ = 0.0;
    pval_ = 0;

    // Alignment Descriptive strings
    contigMatchString_.clear();
    contigAlignedIndexString_.clear();
    contigLostIndexString_.clear();
    opticalMatchString_.clear();
    opticalAlignedIndexString_.clear();
}

File: bin/MatchResult.py
import re
import numpy as np
from scipy.stats import chi2
from math import sqrt
import sys

# Count the number of instances of char c in string s
def countChar(s, c):
    return sum(1 for char in s if char == c)

def computeSd(c):
    sigma = 0.3*sqrt(1000)
    return sigma*sqrt(c)

def printAlignment(mr, fout=sys.stdout):
    mr.printAlignment(fout)

###################################################################
# Define constructors of objects from XML Elements

# The constructor is callable object which processes an xml element
# into an object of some type

# Create an object from the xmlElement text field
class TextConstructor(object):
    def __init__(self, constructor):
        self.constructor = constructor
    def __call__(self, xmlElement):
        return self.constructor(xmlElement.text)

IntField = TextConstructor(int)
FloatField = TextConstructor(float)
StrField = TextConstructor(str)
BoolField = TextConstructor(lambda s: bool(int(s)))

# Constructs a list of elements under the current element with the matching tag
# For example:
# <alignments>
#    <alignment>......</alignment>
#    <alignment>......</alignment>
#    <alignment>......</alignment>
# </alignments>
# Assume ae is Element object for the root node.
# ac = AlignmentConstructor() # This builds an alignment objects
# alc = ListConstructor(ac, 'alignment')
# myAlignments = alc(ae) # This returns a list of 3 alignment objects
class ListConstructor(object):
    def __init__(self, constructor, tag):
        self.constructor = constructor
        self.tag = tag

    def __call__(self, xmlElement):
        children = (e for e in xmlElement.iter() if e.tag==self.tag)
        #return [self.constructor(e) for e in xmlElement.iter(tag=self.tag)]
        return [self.constructor(e) for e in children]

#####################################################################

# A chunk from an optical map
class MapChunk(object):
    def __init__(self, xmlElement):

        # Create list of (key, constructor) pairs
        objList = [ ('start', IntField),
                    ('end', IntField),
                    ('startBp', IntField),
                    ('endBp', IntField),
                    ('frags', ListConstructor(IntField, 'frag'))
                  ]

        # Add attributes to the current object
        self.__dict__.update((key, constructor(xmlElement.find(key))) for key,constructor in objList)
    def lengthBp(self):
        return self.endBp - self.startBp


# The Score of a MatchedChunk
class Score(object):
    def __init__(self, xmlElement):
        objList = [ ('contig', FloatField),
                    ('optical', FloatField),
                    ('sizing', FloatField)
                  ]
        # Add attributes to the current object
        self.__dict__.update((key, constructor(xmlElement.find(key))) for key,constructor in objList)

# A matched chunk
class MatchedChunk(object):
    def __init__(self, xmlElement):
        objList = [ ('contig', MapChunk),
                    ('optical', MapChunk),
                    ('isContigGap', BoolField),
                    ('isBoundaryChunk', BoolField),
                    ('score', Score) ]
        # Add attributes to the current object
        self.__dict__.update((key, constructor(xmlElement.find(key))) for key,constructor in objList)

class MatchResult:
    def __init__(self, *args, **kwargs):
        assert ('xmlElement' in kwargs)
        xmlElement = kwargs['xmlElement']
        self.createFromXmlElement(xmlElement)

    def createFromXmlElement(self, xmlElement):

        # Extract contig information
        contigElement = xmlElement.find('contig');
        self.contigId = contigElement.find('name').text
        self.contigLength = int(contigElement.find('length').text)

        # Define attributes to be read from the xmlElement
        strFields = ['chromosome',
                     'opticalMatchString',
                     'opticalAlignedIndex',
                     'contigMatchString',
                     'scoreString'
                     ]
        intFields = ['cStartIndex', 'cEndIndex', 'cStartBp', 'cEndBp',
                     'cAlignedBases',
                     'opStartIndex',
                     'opEndIndex',
                     'opStartBp',
                     'opEndBp',
                     'opAlignedBases',
                     'totalHits',
                     'totalMisses',
                     'contigHits',
                     'contigMisses',
                     'contigUnalignedBases',
                     'contigUnalignedFrags',
                     'opticalHits',
                     'opticalMisses'
                     ]
        boolFields = ['forward']                     
        floatFields = ['score',
                       'pval',
                       'chi2',
                       'totalMissRate',
                       'contigMissRate',
                       'contigUnalignedBaseRatio',
                       'opticalMissRate'
                       ]

        objList = []
        objList.extend((k, StrField) for k in strFields)
        objList.extend((k, IntField) for k in intFields)
        objList.extend((k, BoolField) for k in boolFields)
        objList.extend((k, FloatField) for k in floatFields)
        objList.append( ('alignment', ListConstructor(MatchedChunk, 'chunk')) )

        # Add attributes to the current object
        self.__dict__.update( (key, constructor(xmlElement.find(key))) for key,constructor in objList )
                       

    # Compute statistics related to the alignment
    # 1/4/2013: This function is now incompatible with the current
    # implementation of the MatchResult class
    def computeStats(self):
        pass
        '''
        self.alignedOpticalLengths = np.array([np.sum(self.opticalFragLengths[frags]) for frags in self.alignedOpticalFragsInd])
        self.alignedOpticalVariance = np.array([np.sum(self.opticalFragSD[frags]**2) for frags in self.alignedOpticalFragsInd])
        self.alignedOpticalVariance = np.array([np.sum(self.opticalFragSD[frags]**2) for frags in self.alignedOpticalFragsInd])
        self.alignedContigVariance = np.array([np.sum(self.contigFragSD[frags]**2) for frags in self.alignedContigFragsInd])
        self.alignedContigLengths = np.array([np.sum(self.contigFragLengths[frags]) for frags in self.alignedContigFragsInd])
        # LMM 4/8: This data set does not have standard deviations reported for the optical fragments.
        # Avoid a divide by zero
        #self.alignedOpticalSD = (self.alignedContigLengths - self.alignedOpticalLengths)/np.sqrt(self.alignedOpticalVariance)
        self.alignedOpticalSD = np.zeros(self.numAlignedFrags)
        self.alignedContigSD = (self.alignedContigLengths - self.alignedOpticalLengths)/np.sqrt(self.alignedContigVariance)

        # Number of bases in a contig fragment that is a one-to-one alignment with optical fragment:
        oneToOneInd = np.array([ind[0] for ind in self.alignedContigFragsInd[1:-1] if ind.shape[0]==1], dtype=int)
        self.basesInOneToOne = np.sum(self.contigFragLengths[oneToOneInd])
        self.basesInOneToOneRatio = float(self.basesInOneToOne)/self.size
        self.medianAlignedSD = -1
        self.meanAlignedSD = -1
        self.minAlignedSD = -1
        self.maxAlignedSD = -1
        self.innerContigLength = -1
        self.innerOpticalLength = -1
        lengths = -1
        self.lengthRatio = -1
        self.alignedSD = self.alignedContigSD
        if self.numAlignedFrags > 2:
            self.medianAlignedSD = np.median(np.abs(self.alignedSD[1:-1]))
            self.meanAlignedSD = np.mean(np.abs(self.alignedSD[1:-1]))
            self.minAlignedSD = np.min(np.abs(self.alignedSD[1:-1]))
            self.maxAlignedSD = np.max(np.abs(self.alignedSD[1:-1]))
            self.innerContigLength = np.sum(self.alignedContigLengths[1:-1]) # Length of the inner aligned fragments
            self.innerOpticalLength = np.sum(self.alignedOpticalLengths[1:-1]) # Length of the inner aligned fragments
            lengths = np.array([self.innerContigLength, self.innerOpticalLength], dtype=float)
            self.lengthRatio = np.min(lengths)/np.max(lengths)
            assert len(self.alignedOpticalLengths) == len(self.alignedOpticalVariance) == len(self.alignedContigLengths)
        '''

    def printAlignment(self, fout=sys.stdout):
        mr = self
        fout.write('%s %i %f\n'%(mr.contigId, mr.contigLength, mr.pval))
        fout.write('%s %i to %i\n'%(mr.chromosome, mr.opStartIndex, mr.opEndIndex))
        cStrings = []
        oStrings = []
        cStrings.append('Contig Frags')
        oStrings.append('Optical Frags')

        # Iterate over the MatchedChunks
        for mc in self.alignment:
            if mc.isContigGap:
                cfragString = ' + '.join(map(str, mc.contig.frags))
                cfragString += ' = ' +  str(np.sum(mc.contig.frags))
                ofragString = 'GAP'
            else:
                cfragString = ' + '.join(map(str, mc.contig.frags))
                cfragString += ' = ' +  str(np.sum(mc.contig.frags))
                ofragString = str(np.sum(mc.optical.frags))
                ofragString += ' = ' + ' + '.join(map(str,mc.optical.frags))
            cStrings.append(cfragString)
            oStrings.append(ofragString)

        cw = max(len(s) for s in cStrings)
        ow = max(len(s) for s in oStrings)
        # Fancy python string formatting:
        for c, o in zip(cStrings, oStrings):
            fout.write('{0:^{width1}} | {1:^{width2}}\n'.format(c, o, width1=cw, width2=ow))
